In [2]:
import pandas as pd

df = pd.read_excel("salesworkload.xlsx", header=1)
df.head()

MonthYear  Time index         Country  StoreID        City  Dept_ID  \
0   10.2016         1.0  United Kingdom  88253.0  London (I)      1.0   
1   10.2016         1.0  United Kingdom  88253.0  London (I)      2.0   
2   10.2016         1.0  United Kingdom  88253.0  London (I)      3.0   
3   10.2016         1.0  United Kingdom  88253.0  London (I)      4.0   
4   10.2016         1.0  United Kingdom  88253.0  London (I)      5.0   

            Dept. Name  HoursOwn  HoursLease  Sales units   Turnover  \
0                  Dry  3184.764         0.0     398560.0  1226244.0   
1               Frozen  1582.941         0.0      82725.0   387810.0   
2                other    47.205         0.0     438400.0   654657.0   
3                 Fish  1623.852         0.0     309425.0   499434.0   
4  Fruits & Vegetables  1759.173         0.0     165515.0   329397.0   

   Customer Area (m2) Opening hours  
0       NaN    953.04        Type A  
1       NaN    720.48        Type A  
2       NaN    966.72        Type A  
3       NaN   1053.36        Type A  
4       NaN   1053.36        Type A

In [4]:
df.shape

(7658, 14)

- **ROWS**: 7,658
- **Columns**: 13

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7658 entries, 0 to 7657
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   MonthYear      7658 non-null   object 
 1   Time index     7650 non-null   float64
 2   Country        7650 non-null   object 
 3   StoreID        7650 non-null   float64
 4   City           7650 non-null   object 
 5   Dept_ID        7650 non-null   float64
 6   Dept. Name     7650 non-null   object 
 7   HoursOwn       7650 non-null   object 
 8   HoursLease     7650 non-null   float64
 9   Sales units    7650 non-null   float64
 10  Turnover       7650 non-null   float64
 11  Customer       0 non-null      float64
 12  Area (m2)      7650 non-null   object 
 13  Opening hours  7650 non-null   object 
dtypes: float64(7), object(7)
memory usage: 837.7+ KB


**TODOs**
- Drop Customers as it is always empty.
- MonthYear and Time index are the same. Whta to do?
- Many fields only have 7650 values. Which means there are 8 empty rows. For some reason. Check that.

In [8]:
df = df.drop('Customer', axis=1)

In [10]:
df.head()

MonthYear  Time index         Country  StoreID        City  Dept_ID  \
0   10.2016         1.0  United Kingdom  88253.0  London (I)      1.0   
1   10.2016         1.0  United Kingdom  88253.0  London (I)      2.0   
2   10.2016         1.0  United Kingdom  88253.0  London (I)      3.0   
3   10.2016         1.0  United Kingdom  88253.0  London (I)      4.0   
4   10.2016         1.0  United Kingdom  88253.0  London (I)      5.0   

            Dept. Name  HoursOwn  HoursLease  Sales units   Turnover  \
0                  Dry  3184.764         0.0     398560.0  1226244.0   
1               Frozen  1582.941         0.0      82725.0   387810.0   
2                other    47.205         0.0     438400.0   654657.0   
3                 Fish  1623.852         0.0     309425.0   499434.0   
4  Fruits & Vegetables  1759.173         0.0     165515.0   329397.0   

  Area (m2) Opening hours  
0    953.04        Type A  
1    720.48        Type A  
2    966.72        Type A  
3   1053.36        Type A  
4   1053.36        Type A

In [14]:
df.isna().sum() # 8 null values in all columns, for some reason. These turned out to be some separators.

MonthYear        0
Time index       8
Country          8
StoreID          8
City             8
Dept_ID          8
Dept. Name       8
HoursOwn         8
HoursLease       8
Sales units      8
Turnover         8
Area (m2)        8
Opening hours    8
dtype: int64

In [18]:
df.describe().T

count          mean           std      min        25%       50%  \
Time index   7650.0  5.000000e+00  2.582158e+00      1.0       3.00       5.0   
StoreID      7650.0  6.199522e+04  2.992458e+04  12227.0   29650.00   75400.5   
Dept_ID      7650.0  9.470588e+00  5.337429e+00      1.0       5.00       9.0   
HoursLease   7650.0  2.203608e+01  1.332995e+02      0.0       0.00       0.0   
Sales units  7650.0  1.076471e+06  1.728113e+06      0.0   54571.25  293230.0   
Turnover     7650.0  3.721393e+06  6.003380e+06      0.0  272679.75  931957.5   

                   75%         max  
Time index         7.0         9.0  
StoreID        87703.0     98422.0  
Dept_ID           14.0        18.0  
HoursLease         0.0      3984.0  
Sales units   917507.5  11242955.0  
Turnover     3264432.0  42717390.0

In [40]:
null_mask = df.isnull().any(axis=1)
null_rows = df[null_mask]

print(null_rows)

Empty DataFrame
Columns: [MonthYear, Time index, Country, StoreID, City, Dept_ID, Dept. Name, HoursOwn, HoursLease, Sales units, Turnover, Area (m2), Opening hours]
Index: []


In [ ]:
# Remove the separators where MonthYear = "- - - -".

In [38]:
df = df[df.MonthYear != " - - - - "]

In [42]:
null_mask = df.isnull().any(axis=1)
null_rows = df[null_mask]

print(null_rows) # No empty rows now

Empty DataFrame
Columns: [MonthYear, Time index, Country, StoreID, City, Dept_ID, Dept. Name, HoursOwn, HoursLease, Sales units, Turnover, Area (m2), Opening hours]
Index: []


In [66]:
pd.options.display.float_format = '{:,.2f}'.format  # Format floats with 2 decimal places and remove all e+XX formats for better clarity.
df.describe().T

count         mean          std       min        25%  \
Time index  7,650.00         5.00         2.58      1.00       3.00   
StoreID     7,650.00    61,995.22    29,924.58 12,227.00  29,650.00   
Dept_ID     7,650.00         9.47         5.34      1.00       5.00   
HoursLease  7,650.00        22.04       133.30      0.00       0.00   
Sales units 7,650.00 1,076,470.71 1,728,113.22      0.00  54,571.25   
Turnover    7,650.00 3,721,392.94 6,003,380.29      0.00 272,679.75   

                   50%          75%           max  
Time index        5.00         7.00          9.00  
StoreID      75,400.50    87,703.00     98,422.00  
Dept_ID           9.00        14.00         18.00  
HoursLease        0.00         0.00      3,984.00  
Sales units 293,230.00   917,507.50 11,242,955.00  
Turnover    931,957.50 3,264,432.00 42,717,390.00

In [74]:
df["Dept. Name"].value_counts()

Dept. Name
Dry                    450
Hardware               450
others                 450
Delivery               450
Customer Services      450
Checkout               450
Admin                  450
Non Food               450
Household              450
Frozen                 450
Clothing               450
Food                   450
Meat                   450
Fruits & Vegetables    450
Fish                   450
other                  450
all                    450
Name: count, dtype: int64

In [76]:
df["Opening hours"].value_counts()

Opening hours
Type A    5661
Type B    1836
Type C     153
Name: count, dtype: int64

In [82]:
# sTypeA -> column["Opening hours"] `Type A`
sTypeC = df[df["Opening hours"] == "Type C"]
sTypeC

MonthYear  Time index Country   StoreID    City  Dept_ID  \
493    10.2016        1.00   Italy 64,983.00  Milano     1.00   
494    10.2016        1.00   Italy 64,983.00  Milano     2.00   
495    10.2016        1.00   Italy 64,983.00  Milano     3.00   
496    10.2016        1.00   Italy 64,983.00  Milano     4.00   
497    10.2016        1.00   Italy 64,983.00  Milano     5.00   
...        ...         ...     ...       ...     ...      ...   
7313   06.2017        9.00   Italy 64,983.00  Milano    12.00   
7314   06.2017        9.00   Italy 64,983.00  Milano    16.00   
7315   06.2017        9.00   Italy 64,983.00  Milano    11.00   
7316   06.2017        9.00   Italy 64,983.00  Milano    17.00   
7317   06.2017        9.00   Italy 64,983.00  Milano    18.00   

               Dept. Name  HoursOwn  HoursLease  Sales units      Turnover  \
493                   Dry  3,669.40        0.00   692,935.00    203,550.00   
494                Frozen  3,571.84        0.00   168,440.00    101,340.00   
495                 other     47.20        0.00   743,810.00  1,262,694.00   
496                  Fish  2,451.51        0.00   488,405.00    869,277.00   
497   Fruits & Vegetables  1,944.85        0.00   246,885.00    636,633.00   
...                   ...       ...         ...          ...           ...   
7313             Checkout  7,357.69       96.00 4,083,590.00 14,278,947.00   
7314    Customer Services  3,080.91       46.00       145.00          0.00   
7315             Delivery  4,928.20        0.00   362,790.00  1,539,012.00   
7316               others  2,665.51        0.00       145.00          0.00   
7317                  all 50,078.21      142.00 4,083,590.00 16,779,936.00   

     Area (m2) Opening hours  
493     816.24        Type C  
494     611.04        Type C  
495     877.80        Type C  
496   1,361.16        Type C  
497   1,003.20        Type C  
...        ...           ...  
7313       #NV        Type C  
7314       #NV        Type C  
7315       #NV        Type C  
7316       #NV        Type C  
7317       #NV        Type C  

[153 rows x 13 columns]

In [86]:
sTypeC.describe().T

count         mean          std       min        25%  \
Time index  153.00         5.00         2.59      1.00       3.00   
StoreID     153.00    64,983.00         0.00 64,983.00  64,983.00   
Dept_ID     153.00         9.47         5.35      1.00       5.00   
HoursLease  153.00        71.41       187.81      0.00       0.00   
Sales units 153.00 1,159,903.79 1,747,041.37    145.00  74,765.00   
Turnover    153.00 3,808,837.14 5,849,753.44      0.00 318,507.00   

                     50%          75%           max  
Time index          5.00         7.00          9.00  
StoreID        64,983.00    64,983.00     64,983.00  
Dept_ID             9.00        14.00         18.00  
HoursLease          0.00        46.00      1,204.00  
Sales units   391,100.00   785,595.00  7,277,805.00  
Turnover    1,105,425.00 3,300,234.00 25,312,311.00

In [92]:
df = df.sort_values(by="Turnover", ascending=False)
df.head()

MonthYear  Time index         Country   StoreID        City  Dept_ID  \
2517   12.2016        3.00          Sweden 81,473.00   Stockholm    18.00   
2513   12.2016        3.00          Sweden 81,473.00   Stockholm    12.00   
2507   12.2016        3.00          Sweden 81,473.00   Stockholm    13.00   
1731   12.2016        3.00  United Kingdom 38,976.00  Manchester    12.00   
1735   12.2016        3.00  United Kingdom 38,976.00  Manchester    18.00   

     Dept. Name  HoursOwn  HoursLease   Sales units      Turnover Area (m2)  \
2517        all 87,338.69       10.00 11,242,955.00 42,717,390.00 38,169.48   
2513   Checkout 14,677.61        0.00 11,242,955.00 42,717,390.00 38,169.48   
2507       Food 41,288.64        0.00 10,208,235.00 35,376,837.00 23,208.12   
1731   Checkout 13,796.45        0.00  9,826,230.00 33,617,103.00 29,521.44   
1735        all 77,409.91        0.00  9,826,230.00 33,617,103.00 29,521.44   

     Opening hours  
2517        Type A  
2513        Type A  
2507        Type A  
1731        Type A  
1735        Type A